File: src/rails/common/GameOptionsSet.java
package rails.common;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.SortedSet;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;

/**
 * A class to store several GameOptions, including values
 */
public class GameOptionsSet {

    private final LinkedHashMap<String, String> optionsToValues= Maps.newLinkedHashMap();
    
    private GameOptionsSet(int nbPlayers) {
        optionsToValues.put(GameOption.NUMBER_OF_PLAYERS, Integer.toString(nbPlayers));
    }
    
    private static GameOptionsSet create(int nbPlayers, List<GameOption> options) {
        GameOptionsSet set = new GameOptionsSet(nbPlayers);
        for (GameOption option:options) {
            set.optionsToValues.put(option.getName(), option.getSelectedValue());
        }
        return set;
    }
    
    public ImmutableMap<String, String> getOptions() {
        return ImmutableMap.copyOf(optionsToValues);
    }
    
    public String get(String option) {
        return optionsToValues.get(option);
    }
    
    public static Builder builder() {
        return new Builder();
    }
    
    public static class Builder {
        private final SortedSet<GameOption> options = Sets.newTreeSet();
        
        private Builder() { }
        
        public void add(GameOption option) {
            options.add(option);
        }
        
        public ImmutableList<GameOption> getOptions() {
            return ImmutableList.copyOf(options);
        }
        
        public GameOptionsSet build(int nbPlayers) {
            return GameOptionsSet.create(nbPlayers, this.getOptions());
        }
    }
    
}

File: dotnet/RailsLib.Net/Net/Common/GameOptionsSet.cs
using Newtonsoft.Json;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;

namespace GameLib.Net.Common
{
    [JsonObject(MemberSerialization.Fields)]
    public class GameOptionsSet
    {
        //private LinkedHashMap<string, string> optionsToValues= new LinkedHashMap<string, string>();
        //private OrderedDictionary optionsToValues = new OrderedDictionary();
        private Dictionary<string, string> optionsToValues = new Dictionary<string, string>();

        public GameOptionsSet() { }

        private GameOptionsSet(int nbPlayers)
        {
            optionsToValues[GameOption.NUMBER_OF_PLAYERS] = nbPlayers.ToString();
        }

        private static GameOptionsSet Create(int nbPlayers, List<GameOption> options)
        {
            GameOptionsSet set = new GameOptionsSet(nbPlayers);
            foreach (GameOption option in options)
            {
                set.optionsToValues[option.Name] = option.SelectedValue;
            }
            return set;
        }

        public Dictionary<string, string> GetOptions()
        {
            return optionsToValues;
            //Dictionary<string, string> ret = new Dictionary<string, string>();
            //IDictionaryEnumerator en = optionsToValues.GetEnumerator();
            //while (en.MoveNext())
            //{
            //    ret[(string)en.Key] = (string)en.Value;
            //}

            //return ret;
        }

        public string Get(string option)
        {
            return (string)optionsToValues[option];
        }

        public static Builder GetBuilder()
        {
            return new Builder();
        }

        public class Builder
        {
            private SortedSet<GameOption> options = new SortedSet<GameOption>();

            // #builder_constructor
            public Builder() { }

            public void Add(GameOption option)
            {
                options.Add(option);
            }

            public List<GameOption> GetOptions()
            {
                return new List<GameOption>(options); //ImmutableList.copyOf(options);
            }

            public GameOptionsSet Build(int nbPlayers)
            {
                return GameOptionsSet.Create(nbPlayers, this.GetOptions());
            }
        }

    }
}
